### DISCUSSION TITLE
pyarmor not working with streamlit

### DISCUSSION THREAD

**Discussion by shamiulshifat** [dashingsoft/pyarmor#1584] [Q&A] [answered]
I have a streamlit app where the architecture is:

/subfolder1
/subfolder2
main.py
start.sh 

Now I'm obfuscating using:

```
pyarmor-7 obfuscate --recursive --restrict=0 --exact--no-cross-protection --no-runtime --exact main.py

```
Then I run "bash start.sh" which is an executable file containing the command 'python -m main.py'
Now I'm getting the error:

```
ModuleNotFoundError: No module named 'pytransform'

raceback:
File "/home/betterdata/Desktop/py3918/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
File "/home/betterdata/Desktop/test-12-11-23/synthetic-static-main/dist/mainpy", line 1, in <module>
    from pytransform import pyarmor_runtime

```
Already checked python version is same, using within python 3.9.18 env, tried many combinations from pyarmor docs, nothing seems to work. 

```

**Comment by georgialive** [comment]
The issue you're experiencing with PyArmor and your Streamlit app seems to be related to the way PyArmor obfuscates the code and the subsequent execution of the obfuscated code. PyArmor obfuscates Python code and typically requires a runtime hook (`pytransform`) to decrypt and execute the obfuscated scripts. Here are some steps and considerations to resolve the issue:

1. **Including PyArmor Runtime**: When you use the `--no-runtime` flag with PyArmor, it doesn't include the PyArmor runtime (`pytransform`) in the obfuscated code. This is likely the cause of the `ModuleNotFoundError: No module named 'pytransform'` error. You might need to include the runtime by removing the `--no-runtime` flag, especially if the obfuscated script is to be run on a different environment or machine.

2. **Correct Use of Flags**: Ensure that the flags you're using with PyArmor are correct and appropriate for your use case. The `--exact` and `--no-cross-protection` flags have specific purposes, and you should use them only if they are necessary for your application.

3. **Script Execution**: In your `start.sh`, you're using `python -m main.py`, which is incorrect syntax for running a Python script. It should be `python main.py` or `python -m module_name` (without `.py`) if `main` is a module. Ensure that the script is executed correctly.

4. **Python Environment Consistency**: Double-check that the Python environment you're using to obfuscate the code is the same as the one you're using to run the obfuscated code. Different Python versions or environments can lead to issues.

5. **Re-Obfuscating After Changes**: If you make changes to the original Python scripts, you need to re-obfuscate them. Any change in the source code requires re-obfuscation.

6. **PyArmor Project Configuration**: Consider using a PyArmor project to manage your obfuscation settings. This can help maintain consistent settings across obfuscations and simplify the process.

7. **Testing in Different Environments**: Test the obfuscated code in an environment similar to your production environment. This helps in identifying and fixing environment-specific issues.

8. **Consult PyArmor Documentation**: Review the PyArmor documentation for guidance on obfuscating Streamlit apps and for any updates or notes on compatibility issues.

9. **Seeking Support**: If you continue to face issues, consider seeking support from the PyArmor community or the tool's maintainers. They might offer specific advice based on their experience with similar cases.

By addressing these points, you should be able to resolve the `ModuleNotFoundError` and get your obfuscated Streamlit app running as expected.

**Reply by shamiulshifat** [reply]
Thanks. 
Sorry my start.sh contains code: python -m streamlit run main.py 

Now I've used command: pyarmor-7 obfuscate --recursive --restrict=0 --no-cross-protection main.py

It's running the streamlit app but when I click on a button on my app, it gives:
```
RuntimeError: Marshal loads failed
Traceback:
File "/home/betterdata/Desktop/py3918/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
File "<<<</home/betterdata/Desktop/test-12-11-23/synthetic-static-main/dist/main.py>>>>", line 3, in <module>
exec(code, module.__dict__)
File "</home/betterdata/Desktop/test-12-11-23/synthetic-static-main/dist/main_podman_st.py>", line 3, in <module>
RuntimeError: Marshal loads failed
2023-11-13 14:24:23.476 Uncaught app exception
Traceback (most recent call last):
File "/home/betterdata/Desktop/py3918/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
exec(code, module.__dict__)
File "<</home/betterdata/Desktop/test-12-11-23/synthetic-static-main/dist/main_podman_st.py>>", line 3, in <module>
RuntimeError: Marshal loads failed
2023-11-13 14:24:31.432 Uncaught app exception
Traceback (most recent call last):
File "/home/betterdata/Desktop/py3918/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
exec(code, module.__dict__)
File "<<</home/betterdata/Desktop/test-12-11-23/synthetic-static-main/dist/main_podman_st.py>>>", line 3, in <module>
RuntimeError: Marshal loads failed
```

**Comment by shamiulshifat** [comment, marked as answer]
Solved the issue by using config:
{
  "version": "2.1",
  "name": "synthetic-static-main",
  "title": "synthetic-static-main",
  "src": ".",
  "is_package": null,
  "manifest": "global-include *.py",
  "entry": "main_podman_st.py",
  "output": "dist",
  "runtime_path": null,
  "restrict_mode": 0,
  "obf_code": 0,
  "obf_mod": 0,
  "wrap_mode": 0,
  "advanced_mode": 0,
  "bootstrap_code": 1,
  "cross_protection": 0,
  "mixins": null,
  "plugins": null,
  "platform": null,
  "package_runtime": 1,
  "enable_suffix": 0,
  "license_file": null,
  "build_time": 1699857259.86041
}

**Comment by happyforx** [comment]
how to solved this problem
